pkg/api/rate_limiter.go

```diff
@@ -0,0 +1,152 @@
+package api
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"dojo-api/pkg/blockchain"
+	"dojo-api/pkg/cache"
+	"dojo-api/utils"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+
+	limiter "github.com/ulule/limiter/v3"
+	sredis "github.com/ulule/limiter/v3/drivers/store/redis"
+)
+
+var (
+	limiters sync.Map
+	once     sync.Once
+)
+
+type RateLimiterKey string
+
+const (
+	WorkerRateLimiterKey    RateLimiterKey = "dojo_worker_api:limiter:worker"
+	WriteTaskRateLimiterKey RateLimiterKey = "dojo_worker_api:limiter:task_write"
+	ReadTaskRateLimiterKey  RateLimiterKey = "dojo_worker_api:limiter:task_read"
+	MetricsRateLimiterKey   RateLimiterKey = "dojo_worker_api:limiter:metrics"
+)
+
+type LimiterConfig struct {
+	key    RateLimiterKey
+	rate   limiter.Rate
+	prefix string
+}
+
+func init() {
+	InitializeLimiters()
+}
+
+func GenerousRateLimiter() gin.HandlerFunc {
+	return getRateLimiterMiddleware(WorkerRateLimiterKey)
+}
+
+func WriteTaskRateLimiter() gin.HandlerFunc {
+	return getRateLimiterMiddleware(WriteTaskRateLimiterKey)
+}
+
+func ReadTaskRateLimiter() gin.HandlerFunc {
+	return getRateLimiterMiddleware(ReadTaskRateLimiterKey)
+}
+
+func MetricsRateLimiter() gin.HandlerFunc {
+	return getRateLimiterMiddleware(MetricsRateLimiterKey)
+}
+
+func InitializeLimiters() {
+	once.Do(func() {
+		cache := cache.GetCacheInstance()
+
+		limiterConfigs := []LimiterConfig{
+			{
+				key:    WorkerRateLimiterKey,
+				rate:   limiter.Rate{Period: 1 * time.Hour, Limit: 3600},
+				prefix: string(WorkerRateLimiterKey),
+			},
+			{
+				key:    WriteTaskRateLimiterKey,
+				rate:   limiter.Rate{Period: 1 * time.Hour, Limit: 60},
+				prefix: string(WriteTaskRateLimiterKey),
+			},
+			{
+				key:    ReadTaskRateLimiterKey,
```

Comment:
Instead of 3600/hr, how about 60/min to avoid the abuse API for 3600 time in continuously.